Attempt to reduce SpreadDamageWarhead CPU overhead

reaperrr opened the pull request:
I'm not sure how large the total positive effect on performance is (assuming there is any), but I think this is worth testing.

This basically does ~two~ three things:
- removes the 2% and 1% stages from default `Falloff`. For the majority of weapons in RA, TD and TS these stages resulted in zero damage or 1 damage point against most actors anyway, while D2k now gets its own, custom Falloffs. This reduces situations where infantry would go prone due to a single damage point, and effectively reduces the default radius that is scanned for victims from 6x to 4x `Spread`, reducing the average number of iterated actors and number of 1-damage-point `Health.InflictDamage` ops.
- `hitActors` now only enumerates actors with `HealthInfo`, which avoids iterating over and performing a pointless `DoImpact` on invulnerable actors without `Health` trait (which might actually yield significant savings when weapons with large area of effect - like nukes - are fired into an area with lots of rocks or invulnerable trees etc., but for now that's just a guess).
- reduces default `TargetExtraSearchRadius` to largest feasible health radius on any shipping mod (1c512). A larger scan radius will only be needed when we increase health radii / health boxes, and only on weapons with relatively low area of effect.

reaperrr committed:
Remove negligible stages from default warhead Falloff

- reduces situations where infantry goes prone even though the impact was rather far away and damage absolutely negligible
- saves a little bit of performance by reducing the total area of effect, resulting in lower average number of calls to Health.InflictDamage at a given Spread value

reaperrr committed:
Skip DoImpact if actor has no Health trait

pchote commented on OpenRA.Mods.Common/Warheads/SpreadDamageWarhead.cs:
```diff
@@ -58,7 +59,7 @@ public override void DoImpact(WPos pos, Actor firedBy, IEnumerable<int> damageMo
 			// This only finds actors where the center is within the search radius,
 			// so we need to search beyond the maximum spread to account for actors with large health radius
-			var hitActors = world.FindActorsInCircle(pos, Range[Range.Length - 1] + TargetExtraSearchRadius);
+			var hitActors = world.FindActorsInCircle(pos, Range[Range.Length - 1] + TargetExtraSearchRadius).Where(a => a.Info.HasTraitInfo<HealthInfo>());
```
This introduces a second trait(info) lookup per actor.  It would be better to keep the one query in the loop, but change the conditional to `if (healthInfo == null) continue;`

reaperrr commented:
Updated.

reaperrr committed:
Reduce default warhead TargetExtraSearchRadius

There aren't any actors in any shipping mod where a health radius larger
than 1c512 would be feasible, so no need to make the default larger than necessary.

RoosterDragon commented:
:+1:

GraionDilach commented:
:+1:

RoosterDragon commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-%28bleed%29/_compare/de6dd63b6782c97e4cf5ebc85e75b623af7cec8c...62a3311c2ea10ccab42bc49d9099c0b1ee9a1674)
